plugins/rule_engines/irods_rule_language/src/configuration.cpp

```diff
@@ -420,20 +422,60 @@ int hash_rules(const std::vector<std::string> &irbs, const int pid, std::string
 
 class make_copy {
 public:
-        make_copy(const std::vector<std::string>& _irbs, const int _pid) : irbs_(_irbs), pid_(_pid) {
-            for(auto const &irb : _irbs) {
-                boost::filesystem::copy_file(get_rule_base_path(irb), get_rule_base_path_copy(irb, _pid));
-            }
+  make_copy(const std::vector<std::string>& _irbs)
+      : irbs_(_irbs)
+  {
+      for (auto const& irb : _irbs) {
+          std::string rulebase;
+          std::ifstream ifs(get_rule_base_path(irb));
+          rule_bases.insert(
+              {irb, std::string({std::istreambuf_iterator<char>(ifs), std::istreambuf_iterator<char>()})});
+      }
+  }
+
+  std::string& get_rulebase(const std::string& irb)
+  {
+      return rule_bases.at(irb);
```

Comment:
This will throw an exception if `irb` doesn't exist in the vector. Are the calls to `get_rulebase` wrapped in a try-catch block?

Reply:
True, originally this was so I could quickly diagnose if something went wrong. There's no try-catch, so I'm thinking maybe instead of throwing here, I log an error and return an empty string? I never hit this case while testing, and it seems somewhat rare to hit (a function requests a rulebase from the copy that doesn't exist?)
What are your thoughts?

Reply:
I don't think you can use an empty string because a rulebase file is allowed to be empty.

I think wrapping the calls to `get_rulebase` is a stronger approach. That gives you more options on how to proceed.

Alternatively, you make the function return `std::optional<std::string>` and check the optional at each call site. This approach would solve the `const` situation too, but at the cost of a copy. _The copy can be worked around via `std::ref` or a pointer._